### ISSUE TITLE
Active Storage + validations

### ISSUE THREAD

**Issue by igorkasyanchuk** [rails/rails#33741]
### Steps to reproduce

Create model like:

```ruby
class Project < ApplicationRecord
  has_one_attached :attachment
end
```

There is no way to validate if the attachment is present or check its content type, etc. If you need - you need to write own `validates :some_method` code.

### Expected behavior

Should be easy, as with other validations. So I've created a gem https://github.com/igorkasyanchuk/active_storage_validations to help me with it.

Gem allows to write something like:

```ruby
validates :attachment, attached: true, content_type: { in: 'application/pdf', message: 'is not a PDF' }
```

If possible I'm ready to prepare it for merge into Rails repo under your guidance. 

### Actual behavior

We have to write own `validates :some_method` code.

### System configuration
Rails 5.2+ (or any where active_storage is added)
Ruby 2.3+

**Comment by BerkhanBerkdemir** [comment]
What do you think to validate size of the item? I didn't see this in your Gem.

**Comment by igorkasyanchuk** [comment]
It's in TODO list) do you want to help with contribution?

**Comment by BerkhanBerkdemir** [comment]
**TL;DR** Yes I do.

Actually, I wanna help under this topic, @igorkasyanchuk, but I'm just new for contributing something.

**Comment by georgeclaghorn** [comment]
Thanks for exploring this, @igorkasyanchuk! Please do open a PR.

**Comment by schweigert** [comment]
:tada: :up:

**Comment by tlbiz** [comment]
:thumbsup: :+1:

**Comment by arefaslani** [comment]
@igorkasyanchuk How to validate file type and size before file uploads to the server?

**Comment by schweigert** [comment]
@arefaslani I'm using the gem created by @igorkasyanchuk and polishing it over time. Currently it does not allow to validate the size, but it is not a difficult validation to do. Just grab the size of the blob. A PR would be welcome ^ - ^

**Comment by aackerman** [comment]
@igorkasyanchuk I think you meant to mention @arefaslani

**Comment by arefaslani** [comment]
@igorkasyanchuk @schweigert I use direct upload to my local disk. How could I validate the upload before the file be uploaded to the server?

**Comment by arefaslani** [comment]
@igorkasyanchuk Thank you. But this only works on my web client. We have a public API too that no one will encounter with this validation on the API. In general, is there a way to validate files uploading to the direct uploads controller? I've created this solution for myself:
```ruby
# app/controller/active_storage/direct_uploads_controller.rb
# frozen_string_literal: true

# Creates a new blob on the server side in anticipation of a direct-to-service upload from the client side.
# When the client-side upload is completed, the signed_blob_id can be submitted as part of the form to reference
# the blob that was created up front.
class ActiveStorage::DirectUploadsController < ActiveStorage::BaseController
  include Authenticatable
  include Errors

  skip_before_action :verify_authenticity_token, :only => [:create]
  before_action :authorize_access_token

  def create
    # FIXME: remove these validations if ActiveStorage validation released
    begin
      validate_blob_args
      blob = ActiveStorage::Blob.create_before_direct_upload!(blob_args)
    rescue ArgumentError => e
      return render json: {
        errors: [title: 'something went wrong', detail: e]
      }, status: :unprocessable_entity
    end
    render json: direct_upload_json(blob)
  end

  private
    def blob_args
      params.require(:blob).permit(:filename, :byte_size, :checksum, :content_type, :metadata).to_h.symbolize_keys
    end

    def direct_upload_json(blob)
      blob.as_json(root: false, methods: :signed_id).merge(direct_upload: {
        url: blob.service_url_for_direct_upload,
        headers: blob.service_headers_for_direct_upload
      })
    end

    def validate_blob_args
      # can not upload files greater than 10 megabytes
      if blob_args[:byte_size] > 10.megabytes
        raise ArgumentError.new('file is too large')
      end

      # can not upload files less than 1 kilobytes
      if blob_args[:byte_size] < 1
        raise ArgumentError.new('file is too small')
      end

      # only jpeg or png files can be uploaded
      if blob_args[:content_type] &&
          !['image/jpeg', 'image/png'].include?(blob_args[:content_type])
        raise ArgumentError.new('Only jpeg or png files can be uploaded')
      end
    end
end
```
But it's not flexible enough. Maybe I want to upload videos for another resource.

**Comment by igorkasyanchuk** [comment]
@arefaslani to be honest I've never tried direct uploads with ActiveStorage.
But what I can see, you can do following. https://gist.github.com/igorkasyanchuk/88d83d06f579108c50f761da91786af1

**Comment by schweigert** [comment]
What do you think of implementing this controller in your gem @igorkasyanchuk ?

**Comment by igorkasyanchuk** [comment]
if you want :) please implement if you want

**Comment by Uysim** [comment]
🎉 🆙

**Comment by Uysim** [comment]
I wish to see this in Active Storage

**Comment by paulmthiebauth** [comment]
Does this gem prevent AS::Attachment and AS::Blob records from being created in the event an upload doesnt pass validations?

**Comment by pastullo** [comment]
I rushed to upgrade to 5.2 and remove Paperclip and then i realized that current support for AS is very limited. Why did they deprecated Paperclip so quickly if there is still no native validation support?

**Comment by MohamedBasaleh** [comment]
Thanks @igorkasyanchuk 
:+1:

**Comment by connorshea** [comment]
I hope #34842 – or something like it – gets merged 🙏

**Comment by henrik** [comment]
Found this via Google. I'm not willing to try the plugin mentioned above since its README suggests it's not tested broadly.

The below is what I'm doing as a workaround for now. It seems a decent solution for my needs, but yours may vary.

After validation, I check for attachments and send the user back to the form. So I don't prevent it from being created (since this seemed to be the tricky part of the problem), but I immediately tell the user to fix it:

``` ruby
if record.update_attributes(my_attributes)
  if !item.images?  # "images" is the name of my attachments
    redirect_to edit_item_path(item), alert: "You need to provide images!"
  else
    # success!
  end
 
else
  # handle validation errors
end
```

I then combine this with hiding items-without-image from the public part of the web site, so they don't show up until they're fixed. Example scope:

``` ruby
scope :has_images, -> { where("(SELECT 1 FROM active_storage_attachments AS asa WHERE asa.record_id = items.id AND asa.record_type = 'Item' AND name = 'raw_images' LIMIT 1) = 1") }
```

**Comment by connorshea** [comment]
This is still relevant.

**Comment by rafaelfranca** [comment]
This is more a feature request, right? We don't take feature requests in the issue tracker, so I'm going to close. We are open for PRs though like George mentioned.

**Comment by connorshea** [comment]
@rafaelfranca Make sense 👍 

FWIW a PR for this issue (#35390) has been waiting for review for a while. You're probably busy with Rails 6 atm, but if we could get any eyes on that PR it'd be greatly appreciated :)

**Comment by shiloh-sandbag83** [comment]
reflect on this +1 answer
